### ISSUE TITLE
Pool upload on VT restart fails

### ISSUE THREAD

**Issue by GwnDaan** [ad3154/ISO11783-CAN-Stack#64]
Hi! Awesome work on getting this all to work. I didn't expect it to work the first time, but after running the commands, VT showed up smoothly. So great stuff! There are a few minor issues I came across, which I will put here.

When you restart the physical terminal while the pool is loaded in and [visible on the terminal](https://user-images.githubusercontent.com/29043784/206693455-7b9cf727-4bfc-4c53-bf46-e64aefc1aa1d.jpg), you get an [error message](https://user-images.githubusercontent.com/29043784/206693794-f5f2232e-ffd0-4602-ab3c-5a8aeb4065cb.jpg) from the terminal. 

Translated error message: "Error uploading control mask. Disconnect and reconnect the implement. Contact the implement manufacturer if the error persists."

Though upon restarting the example application, the object pool loads fine even after the above error.

**Diagnostics:**
Software: Running [latest](https://github.com/ad3154/ISO11783-CAN-Stack/tree/aa144ad39ce8a430df38ef946d91af44087dc289) unchanged VT example
Hardware: Linux 5.15.74-v7+ #1595 SMP Wed Oct 26 11:03:05 BST 2022 armv7l GNU/Linux
Furthermore I can provide a candump or something similar if that helps. I won't be able to link it here due to the logs containing sensitive data.

**Comment by ad3154** [comment]
It sounds like it's probably also not reacting properly to VT server status message timeouts? I will have to investigate that as well.

**Comment by ad3154** [comment]
@GwnDaan I think I've addressed this issue in this PR: #73 
If you have some time to re-test this your feedback is most appreciated!

**Comment by GwnDaan** [comment]
It seems to be working most of the times now, but I was able reproduce the original upload fail once. The mask never got uploaded to the UT. Here is the log
```
pi@tractorcan:~/ISO11783-CAN-Stack $ ./build/examples/vt_version_3_object_pool/VT3ExampleTarget
Loaded object pool from VT3TestPool.iop
[NM]: New Control function 247
[NM]: A Partner Has Claimed 131
[NM]: New Control function 240
[NM]: New Control function 128
[NM]: New Control function 140
[ETP]: New ETP Session. Dest: 131
[ETP]: Session Closed
[NM]: A Partner Has Claimed 131
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[ETP]: New ETP Session. Dest: 131
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[ETP]: Abort CTS while in data session
[ETP]: Received an abort for an session with PGN: 59136
[ETP]: Session Closed
[ETP]: Received an unexpected or invalid data transfer frame
[ETP]: Received an unexpected or invalid data transfer frame
[ETP]: Received an unexpected or invalid data transfer frame
[ETP]: Received an unexpected or invalid data transfer frame
[ETP]: Received an unexpected or invalid data transfer frame
[ETP]: Received an unexpected or invalid data transfer frame
[ETP]: Received an unexpected or invalid data transfer frame
[ETP]: Received an unexpected or invalid data transfer frame
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Abort EOM without matching session
[ETP]: Session Closed
```

Then there is the other times, where the logs seem weird but eventually the pool always managed to successfully upload, unlike in the log above. Note that I restarted the terminal multiple times in the log below and at different times during/after the upload.
```
pi@tractorcan:~/ISO11783-CAN-Stack $ ./build/examples/vt_version_3_object_pool/VT3ExampleTarget
Loaded object pool from VT3TestPool.iop
[NM]: New Control function 140
[NM]: A Partner Has Claimed 131
[NM]: New Control function 247
[NM]: New Control function 240
[NM]: New Control function 128
[ETP]: New ETP Session. Dest: 131
[ETP]: Session Closed
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[ETP]: New ETP Session. Dest: 131
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Abort EOM without matching session
[ETP]: Session Closed
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Abort EOM without matching session
[ETP]: Session Closed
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Abort EOM without matching session
[ETP]: Session Closed
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[ETP]: New ETP Session. Dest: 131
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Abort EOM without matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[ETP]: Aborting session, T2-3 timeout reached while waiting for CTS
[ETP]: Session Closed
[VT]: An object pool failed to upload. Resetting connection to VT.
[NM]: A Partner Has Claimed 131
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[ETP]: New ETP Session. Dest: 131
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Abort EOM without matching session
[ETP]: Session Closed
[NM]: A Partner Has Claimed 131
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Abort EOM without matching session
[ETP]: Aborting session, T2-3 timeout reached while waiting for CTS
[ETP]: Session Closed
[VT]: An object pool failed to upload. Resetting connection to VT.
[NM]: A Partner Has Claimed 131
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[ETP]: New ETP Session. Dest: 131
[TP]: Received an abort
[ETP]: Aborting session, T2-3 timeout reached while waiting for CTS
[ETP]: Session Closed
[VT]: An object pool failed to upload. Resetting connection to VT.
[NM]: A Partner Has Claimed 131
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[ETP]: New ETP Session. Dest: 131
[TP]: Received an abort
[TP]: Abort CTS while in data session
[TP]: Session Closed
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Received an abort
[TP]: Session Closed
[TP]: Abort CTS With no matching session
[TP]: Invalid BAM TP Data Received
[TP]: Invalid BAM TP Data Received
[TP]: Abort EOM without matching session
```
Lmk if there's anything I can do to help

**Comment by ad3154** [comment]
Hm, it looks like the VT sent us an ETP clear to send while we were already transmitting? That is strange. I will look into these other log outputs. If you happen to catch this issue again, a CAN trace would be very helpful. You can capture one by running `candump -l can0` in a separate terminal before running the program.

**Comment by GwnDaan** [comment]
Okay sounds good I'll try to reproduce it tomorrow

**Comment by ad3154** [comment]
Many of the other [TP] logs are another bug 😢 I will open a separate issue for those.

**Comment by ad3154** [comment]
I've resolved the bulk of that weird log stuff in #78 so if you do try and reproduce, please make sure to pull the latest version of the `bugfix/vt-reconnection` branch, as I have rebased it on main.

**Comment by GwnDaan** [comment]
I think this issue is solved. I didn't get any failed uploads or weird logs when testing today. Here is the output of one of the many tests:
```
Loaded object pool from VT3TestPool.iop
[NM]: New Control function 247
[NM]: A Partner Has Claimed 131
[NM]: New Control function 140
[NM]: New Control function 240
[NM]: New Control function 128
[ETP]: New ETP Session. Dest: 131
[ETP]: Session Closed
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[ETP]: Session Closed
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[ETP]: Aborting session, T2-3 timeout reached while waiting for CTS
[ETP]: Session Closed
[VT]: An object pool failed to upload. Resetting connection to VT.
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[ETP]: Aborting session, T2-3 timeout reached while waiting for CTS
[ETP]: Session Closed
[VT]: An object pool failed to upload. Resetting connection to VT.
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[ETP]: Aborting session, T2-3 timeout reached while waiting for CTS
[ETP]: Session Closed
[VT]: An object pool failed to upload. Resetting connection to VT.
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[ETP]: Session Closed
[VT]: Status Timeout
[NM]: A Partner Has Claimed 131
[ETP]: New ETP Session. Dest: 131
[ETP]: Session Closed
```
